src/sentry/event_manager.py

```diff
@@ -1620,11 +1622,37 @@ def _save_aggregate(
         # _save_aggregate had races around group creation which made this race
         # more user visible. For more context, see 84c6f75a and d0e22787, as
         # well as GH-5085.
-        GroupHash.objects.filter(id__in=[h.id for h in new_hashes]).exclude(
-            state=GroupHash.State.LOCKED_IN_MIGRATION
-        ).update(group=group)
-
-    is_regression = _process_existing_aggregate(
+        with transaction.atomic(using="default"):
+            cur = connections["default"].cursor()
+            statement_timeout = None
+            try:
+                # This is a hacky way to prevent contention on the group hash locks.  Workaround
+                # for INC-377 related to how we already have a statement timeout for project
+                # counters.  See SENTRY_PROJECT_COUNTER_STATEMENT_TIMEOUT for related issue.
+                if settings.SENTRY_GROUP_HASH_ASSOCIATION_STATEMENT_TIMEOUT is not None:
+                    cur.execute("show statement_timeout")
+                    statement_timeout = cur.fetchone()[0]
+                    cur.execute(
+                        "set local statement_timeout = %s",
+                        [settings.SENTRY_GROUP_HASH_ASSOCIATION_STATEMENT_TIMEOUT],
+                    )
+                GroupHash.objects.filter(id__in=[h.id for h in new_hashes]).exclude(
+                    state=GroupHash.State.LOCKED_IN_MIGRATION
+                ).update(group=group)
+            except OperationalError:
+                metrics.incr(
+                    "group_hash_association.timeout",
+                    tags={"platform": event.platform or "unknown"},
+                )
+                sentry_sdk.capture_message("group_hash_association.timeout")
+                # Better not regress on a hash we did not manage to associate
+                skip_regression_check = True
```

Comment:
I don't feel great about the behavior of this on timeout, taking suggestions.

Reply:
I assume we don't want to retry at all, since we'd probably just increase lock contention?